File: clients/go/http.go
package wavehouse

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"math/rand/v2"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

var errAborted = &Error{Status: 0, Code: "ABORTED", Message: "Request aborted", Retryable: false}

// maxRetryAfter caps server-supplied Retry-After delays so a hostile or
// misconfigured server can't park the calling goroutine for hours.
const maxRetryAfter = 30 * time.Second

// httpContext carries per-client state needed by every request.
type httpContext struct {
	baseURL    string
	auth       func(ctx context.Context) (string, error)
	maxRetries int
	httpClient *http.Client
	headers    map[string]string
}

// applyConfiguredHeaders writes the client's configured headers onto a request.
// Call it *before* the SDK sets its own headers: Set replaces, so whatever the
// SDK writes afterwards wins a collision. http.Header canonicalizes names, so
// "x-tenant" and "X-Tenant" are the same entry.
func applyConfiguredHeaders(h http.Header, configured map[string]string) {
	for k, v := range configured {
		h.Set(k, v)
	}
}

// requestOptions describes a single HTTP request.
type requestOptions struct {
	method      string
	path        string
	body        any    // JSON-serialized if non-nil
	rawBody     string // sent verbatim if non-empty (takes precedence over body)
	contentType string // overrides Content-Type (default "application/json")
	params      url.Values
}

// doRequest is the internal fetch wrapper with auth, retry, and backoff.
// It decodes the response body into dst (unless dst is nil).
func doRequest(ctx context.Context, hctx httpContext, opts requestOptions, dst any) error {
	reqURL := buildURL(hctx.baseURL, opts.path, opts.params)
	ct := opts.contentType
	if ct == "" {
		ct = "application/json"
	}

	// Serialize body once so every retry sends identical bytes.
	var bodyBytes []byte
	if opts.rawBody != "" {
		bodyBytes = []byte(opts.rawBody)
	} else if opts.body != nil {
		var err error
		bodyBytes, err = json.Marshal(opts.body)
		if err != nil {
			return fmt.Errorf("wavehouse: marshal request body: %w", err)
		}
	}

	// Resolve auth once per request (not per attempt).
	var authHeader string
	if hctx.auth != nil {
		token, err := hctx.auth(ctx)
		if err != nil {
			return fmt.Errorf("wavehouse: auth provider: %w", err)
		}
		if token != "" {
			authHeader = "Bearer " + token
		}
	}

	var lastErr error
	maxAttempts := hctx.maxRetries + 1

	// Retries all methods including POST. For /v1/ingest, at-least-once delivery
	// is the documented contract; dedup is the server-side safety net.
	for attempt := range maxAttempts {
		var bodyReader io.Reader
		if bodyBytes != nil {
			bodyReader = bytes.NewReader(bodyBytes)
		}

		req, err := http.NewRequestWithContext(ctx, opts.method, reqURL, bodyReader)
		if err != nil {
			return fmt.Errorf("wavehouse: build request: %w", err)
		}
		applyConfiguredHeaders(req.Header, hctx.headers)
		req.Header.Set("Content-Type", ct)
		req.Header.Set("Accept", "application/json")
		if authHeader != "" {
			req.Header.Set("Authorization", authHeader)
		}

		res, err := hctx.httpClient.Do(req)
		if err != nil {
			// Context cancellation — return immediately, no retry.
			if ctx.Err() != nil {
				return errAborted
			}
			lastErr = networkError(err)
			if attempt < maxAttempts-1 {
				if sleepErr := sleepWithContext(ctx, backoff(attempt)); sleepErr != nil {
					return errAborted
				}
			}
			continue
		}

		if res.StatusCode >= 200 && res.StatusCode < 300 {
			defer func() { _ = res.Body.Close() }()
			if dst == nil {
				_, _ = io.Copy(io.Discard, res.Body)
				return nil
			}
			raw, readErr := io.ReadAll(res.Body)
			if readErr != nil {
				return networkError(readErr)
			}
			if len(raw) == 0 {
				return nil
			}
			if err := json.Unmarshal(raw, dst); err != nil {
				return &Error{
					Status:    0,
					Code:      "NETWORK_ERROR",
					Message:   fmt.Errorf("decode response: %w", err).Error(),
					Retryable: false,
				}
			}
			return nil
		}

		apiErr := parseErrorResponse(res)
		_ = res.Body.Close()

		// 503/429 with Retry-After: wait the specified duration (capped).
		if res.StatusCode == http.StatusServiceUnavailable || res.StatusCode == http.StatusTooManyRequests {
			if ra := res.Header.Get("Retry-After"); ra != "" && attempt < maxAttempts-1 {
				if sleepErr := sleepWithContext(ctx, retryAfterDelay(ra, attempt)); sleepErr != nil {
					return errAborted
				}
				lastErr = apiErr
				continue
			}
		}

		// Retryable server errors (5xx).
		if apiErr.Retryable && attempt < maxAttempts-1 {
			if sleepErr := sleepWithContext(ctx, backoff(attempt)); sleepErr != nil {
				return errAborted
			}
			lastErr = apiErr
			continue
		}

		return apiErr
	}

	return lastErr
}

func buildURL(base, path string, params url.Values) string {
	u := base + path
	if len(params) > 0 {
		u += "?" + params.Encode()
	}
	return u
}

// retryAfterDelay resolves a Retry-After header (delta-seconds or HTTP-date)
// into a wait, clamped to maxRetryAfter. An unparseable header falls back to
// the ordinary backoff for this attempt, not the maximum.
func retryAfterDelay(ra string, attempt int) time.Duration {
	delay := backoff(attempt)
	if secs, err := strconv.Atoi(ra); err == nil && secs > 0 {
		// Compare before converting: time.Duration(secs) * time.Second wraps
		// negative past ~9.2e9 seconds, and min() below would then pick the
		// negative value, firing the retry timer instantly.
		if secs > int(maxRetryAfter/time.Second) {
			return maxRetryAfter
		}
		delay = time.Duration(secs) * time.Second
	} else if parsed, err := http.ParseTime(ra); err == nil {
		if d := time.Until(parsed); d > 0 {
			delay = d
		}
	}
	return min(delay, maxRetryAfter)
}

func backoff(attempt int) time.Duration {
	ms := 1000 * math.Pow(2, float64(attempt))
	// ±20% jitter so clients failing at the same moment don't retry in
	// lockstep; capped after jitter so the documented 30s max holds.
	ms *= 0.8 + 0.4*rand.Float64() //nolint:gosec // retry jitter, not cryptographic
	return time.Duration(min(ms, 30000)) * time.Millisecond
}

func sleepWithContext(ctx context.Context, d time.Duration) error {
	if ctx.Err() != nil {
		return ctx.Err()
	}
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-t.C:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

File: clients/ts/src/http.ts
import { networkError, parseErrorResponse } from "./errors.js";
import type { HttpContext, WaveHouseError } from "./types.js";
import { resolveURL } from "./url.js";

interface RequestSpec {
  method: string;
  path: string;
  body?: unknown;
  /**
   * Pre-serialized request body, sent verbatim (no `JSON.stringify`) — e.g.
   * NDJSON text. Takes precedence over `body`. Must be re-sendable so retries
   * work, hence a string rather than a stream.
   */
  rawBody?: string;
  /** Override the request Content-Type (default `application/json`). */
  contentType?: string;
  params?: Record<string, string>;
  signal?: AbortSignal;
}

/** @internal */
export interface HttpResult<T> {
  data: T | null;
  error: WaveHouseError | null;
  headers: Headers;
}

/**
 * Merge configured headers underneath the SDK's own, matching names
 * case-insensitively so a caller's `authorization` can't sit alongside our
 * `Authorization` and let the server pick. Canonical casing wins, and values
 * replace rather than append — a header joined instead of replaced is a known
 * way to produce `Content-Type: application/json, image/png`.
 *
 * Shared with the SSE transport so both paths resolve `headers` the same way,
 * rather than growing a second precedence story.
 *
 * @internal
 */
export function mergeHeaders(
  base: Record<string, string>,
  extra: Record<string, string> | undefined,
): Record<string, string> {
  if (!extra) return base;
  const sdkNames = new Set(Object.keys(base).map((k) => k.toLowerCase()));
  const merged = { ...base };
  // Spelling each configured name was last stored under, so two entries
  // differing only in case replace each other here rather than surviving as
  // separate keys for `Headers` to comma-join at fetch time.
  const configured = new Map<string, string>();
  for (const [name, value] of Object.entries(extra)) {
    const lower = name.toLowerCase();
    // Skip rather than overwrite: `base` is the SDK's own, which outranks.
    if (sdkNames.has(lower)) continue;
    const prior = configured.get(lower);
    if (prior !== undefined) delete merged[prior];
    merged[name] = value;
    configured.set(lower, name);
  }
  return merged;
}

/** The documented shape for a cancelled request: a Result, never a throw. */
function abortedResult<T>(): HttpResult<T> {
  return {
    data: null,
    error: { status: 0, code: "ABORTED", message: "Request aborted", retryable: false },
    headers: new Headers(),
  };
}

/**
 * Internal fetch wrapper with auth injection, retry, backoff, and Retry-After.
 * @internal
 */
export async function request<T>(ctx: HttpContext, opts: RequestSpec): Promise<HttpResult<T>> {
  const url = resolveURL(ctx.baseURL, opts.path, opts.params).toString();
  const headers: Record<string, string> = {
    "Content-Type": opts.contentType ?? "application/json",
    Accept: "application/json",
  };

  // Serialize once so every retry attempt re-sends an identical body.
  const requestBody: string | undefined =
    opts.rawBody !== undefined
      ? opts.rawBody
      : opts.body !== undefined
        ? JSON.stringify(opts.body)
        : undefined;

  if (ctx.auth) {
    const token = await ctx.auth();
    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }
  }

  // After auth, so `auth` keeps ownership of Authorization, and after the
  // Content-Type/Accept this request needs.
  const finalHeaders = mergeHeaders(headers, ctx.options.headers);

  let lastError: WaveHouseError | null = null;
  const maxAttempts = ctx.options.maxRetries + 1;

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      // Configured RequestInit first, so the fields the SDK controls overwrite
      // it — a supplied `method` or `body` would otherwise corrupt the request.
      const init: RequestInit = {
        ...ctx.options.fetchOptions,
        method: opts.method,
        headers: finalHeaders,
        body: requestBody,
        signal: opts.signal,
      };
      // Call the global directly when no override is configured, rather than
      // capturing it — a detached `fetch` reference is not universally safe to
      // invoke, and late binding is what lets tests swap `globalThis.fetch`.
      const doFetch = ctx.options.fetch;
      const res = doFetch ? await doFetch(url, init) : await fetch(url, init);

      if (res.ok) {
        const text = await res.text();
        const data = text ? (JSON.parse(text) as T) : (undefined as T);
        return { data, error: null, headers: res.headers };
      }

      const error = await parseErrorResponse(res);

      // 503 with Retry-After: wait the specified duration
      if (res.status === 503) {
        const retryAfter = res.headers.get("Retry-After");
        if (retryAfter && attempt < maxAttempts - 1) {
          const delay = parseInt(retryAfter, 10) * 1000 || 30_000;
          await sleep(delay, opts.signal);
          lastError = error;
          continue;
        }
      }

      // Retryable server errors (5xx)
      if (error.retryable && attempt < maxAttempts - 1) {
        await sleep(backoff(attempt), opts.signal);
        lastError = error;
        continue;
      }

      return { data: null, error, headers: res.headers };
    } catch (e) {
      // AbortError — return immediately, no retry
      if (e instanceof DOMException && e.name === "AbortError") {
        return abortedResult<T>();
      }

      lastError = networkError(e);
      if (attempt < maxAttempts - 1) {
        // This sleep is the one that runs inside the catch, so unlike the two
        // in the try above, an abort raised *by* the backoff would escape
        // `request()` as a raw DOMException instead of the documented ABORTED
        // result — and no caller wraps this, so it would reach the consumer as
        // an unhandled rejection.
        try {
          await sleep(backoff(attempt), opts.signal);
        } catch (abort) {
          if (abort instanceof DOMException && abort.name === "AbortError") {
            return abortedResult<T>();
          }
          throw abort;
        }
      }
    }
  }

  return { data: null, error: lastError!, headers: new Headers() };
}

function backoff(attempt: number): number {
  return Math.min(1000 * 2 ** attempt, 30_000);
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(new DOMException("Aborted", "AbortError"));
      return;
    }
    const timer = setTimeout(resolve, ms);
    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(timer);
        reject(new DOMException("Aborted", "AbortError"));
      },
      { once: true },
    );
  });
}
